### DISCUSSION TITLE
Error switching TypeScript API endpoint from Edge runtime to Node.js runtime

### DISCUSSION THREAD

**Discussion by jeffsmohan** [vercel/community#6252] [Help]
### Summary

I switched one of my serverless functions from the Edge runtime to a Node.js runtime since it now relies on some Node libraries not available on Edge. However, once I switch environments, the API endpoint starts failing with the following error: 

```
ReferenceError: exports is not defined in ES module scope
This file is being treated as an ES module because it has a '.js' file extension and '/var/task/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
```

The function in question doesn't use `exports` (it uses `import` and `export default`). I've tried changing my tsconfig settings every way I can think of. Nothing's made any progress. 

I must be missing something obvious — what do I need to change to get TypeScript working on Node serverless functions? Thanks for any help!

### Example

https://github.com/jeffsmohan/vercel-issue-repro/blob/main/api/repro.ts

### Steps to Reproduce

Here's a minimal reproduction of the issue, looking at a basic hello world API endpoint: https://github.com/jeffsmohan/vercel-issue-repro/blob/main/api/repro.ts

That repo started from the default vite example project (from running `vercel init`), then added the `api/repro.ts` file. When accessing `{{domain}}/api/repro` you'll trigger the exact same error in the logs I've been seeing: 

```
ReferenceError: exports is not defined in ES module scope
This file is being treated as an ES module because it has a '.js' file extension and '/var/task/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
    at file:///var/task/api/repro.js:2:23
    at ModuleJob.run (node:internal/modules/esm/module_job:195:25)
    at ModuleLoader.import (node:internal/modules/esm/loader:336:24)
    at importModuleDynamicallyWrapper (node:internal/vm/module:429:15)
    at d (/opt/node-bridge/bridge-server-CNJTFAIF.js:4:3663)
INIT_REPORT Init Duration: 350.42 ms	Phase: invoke	Status: error	Error Type: Runtime.ExitError
Error: Runtime exited with error: exit status 1
```

**Comment by amyegan** [comment]
Hey @jeffsmohan. Does it succeed if you remove `"type": "module",` from `package.json`?

**Reply by jeffsmohan** [reply]
Thanks for the idea! I tried this, and it does at first blush get the deployed API endpoint working. However, it causes the endpoint to fail in development (via `vercel dev`) now. 

I'll keep poking at this to see if I can figure out how to properly reconcile the different behavior, but if you've got any more insight, I'm all ears!

**Reply by jeffsmohan** [reply]
I've tested a 2x2x2x2 matrix of factors — JS vs TS, Edge vs Node runtime, dev vs prod, and package.json type: module vs not — and here are the results: 

| Lang | Runtime | type:module | **Dev** | **Prod** |
|--------|--------|--------|--------|--------|
| JS | Edge | **type:module** | ✅ | ✅ |
| JS | Edge | –                    | ✅ | ✅ |
| JS | **Node** | **type:module** | ✅ | ✅ |
| JS | **Node** | —                   | ✅ | ✅ |
| **TS** | Edge | **type:module** | ✅ | ✅ |
| **TS** | Edge | —                   | ✅ | ✅ |
| **TS** | **Node** | **type:module** | ✅ | ❌ | 
| **TS** | **Node** | —                  | ❌ | ✅ | 

Error details:
- TS/Node/type:module/prod: `ReferenceError: exports is not defined in ES module scope`
- TS/Node/—/dev: `Error: Unexpected token 'export'`

To test the above, I added four endpoints (edge-ts, edge-js, node-ts, node-js) to my repro GitHub repo (https://github.com/jeffsmohan/vercel-issue-repro/tree/22286f18d52e60663248ec479f45e8d7e5df0dee/api) and ran that in dev then prod; then removed `"type": "module"` from my `package.json` and re-ran the test APIs in dev and prod.
